### ISSUE TITLE
Crashes in background, not sure why.

### ISSUE THREAD

**Issue by wildwestrom** [Tobi823/ffupdater#78]
Android 11
CalyxOS 2.8.0
FFUpdater 74.3.5

```stacktrace
de.marmaro.krt.ffupdater.app.impl.exceptions.ApiNetworkException at ApiConsumer.kt:69
de.marmaro.krt.ffupdater.app.impl.exceptions.ApiNetworkException

de.marmaro.krt.ffupdater.app.impl.exceptions.ApiNetworkException: Failed to consume API. Previous exceptions: [Unable to resolve host "firefox-ci-tc.services.mozilla.com": No address associated with hostname; Unable to resolve host "firefox-ci-tc.services.mozilla.com": No address associated with hostname; Unable to resolve host "firefox-ci-tc.services.mozilla.com": No address associated with hostname; Unable to resolve host "firefox-ci-tc.services.mozilla.com": No address associated with hostname; Unable to resolve host "firefox-ci-tc.services.mozilla.com": No address associated with hostname].
	at de.marmaro.krt.ffupdater.app.impl.fetch.ApiConsumer.readNetworkResourceWithRetries(ApiConsumer.kt:69)
	at de.marmaro.krt.ffupdater.app.impl.fetch.ApiConsumer.access$readNetworkResourceWithRetries(ApiConsumer.kt:20)
	at de.marmaro.krt.ffupdater.app.impl.fetch.ApiConsumer$readNetworkResourceWithRetries$1.invokeSuspend(Unknown Source:15)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.net.UnknownHostException: Unable to resolve host "firefox-ci-tc.services.mozilla.com": No address associated with hostname
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:156)
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:103)
	at java.net.InetAddress.getAllByName(InetAddress.java:1152)
	at com.android.okhttp.Dns$1.lookup(Dns.java:41)
	at com.android.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:178)
	at com.android.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:144)
	at com.android.okhttp.internal.http.RouteSelector.next(RouteSelector.java:86)
	at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:176)
	at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:128)
	at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:97)
	at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:289)
	at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:232)
	at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:465)
	at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:411)
	at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream(HttpURLConnectionImpl.java:248)
	at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getInputStream(DelegatingHttpsURLConnection.java:211)
	at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:30)
	at de.marmaro.krt.ffupdater.app.impl.fetch.ApiConsumer.readNetworkResource(ApiConsumer.kt:80)
	at de.marmaro.krt.ffupdater.app.impl.fetch.ApiConsumer.readNetworkResourceWithRetries(ApiConsumer.kt:54)
	... 8 more
Caused by: android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname)
	at libcore.io.Linux.android_getaddrinfo(Native Method)
	at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:73)
	at libcore.io.BlockGuardOs.android_getaddrinfo(BlockGuardOs.java:202)
	at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:73)
	at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:135)
	... 26 more


Android Version: 30
Device Manufacturer: Google
Device Model: Pixel 4a
```

**Comment by Tobi823** [comment]
The real problem is that the background update check is run even though there is no internet connection. But I don't know why exactly.

Are you using any VPN or setting that communicates the internet availability incorrectly to the operating system? Maybe with this info I can recreate the error myself.

**Comment by Tobi823** [comment]
I have to correct myself:
The real problem is that the background update check is unable to resolve "firefox-ci-tc.services.mozilla.com" (DNS).

What DNS server are you using?

**Comment by Tobi823** [comment]
Maybe your are using AdGuard from https://adguard.com/en/adguard-android/overview.html ?

**Comment by wildwestrom** [comment]
> The real problem is that the background update check is run even though there is no internet connection. But I don't know why exactly.
> 
> Are you using any VPN or setting that communicates the internet availability incorrectly to the operating system? Maybe with this info I can recreate the error myself.

Yes, I am using a VPN.

**Comment by Tobi823** [comment]
It's now fixed in 74.4.3 by improving the internet check by:
- checking the capabilities of the current active network ( NET_CAPABILITY_INTERNET and NET_CAPABILITY_VALIDATED)
- checking if api.github.com and firefox-ci-tc.services.mozilla.com can be resolved (DNS). 
- I could also ping google.com, but it should be as privacy friendly as possible. That's why I don't ping a google address.

You can either wait for the offical F-Droid release, install the APK file from GitHub/GitLab or try my personal F-Droid repository (see https://github.com/Tobi823/ffupdater#f-droid-repository).

I hope I don't upset the mighty GitHub gods by using a GitHub repository as an F-Droid repository :smile:
